### ISSUE TITLE
Problème de cluster

### ISSUE THREAD

**Issue by fefe78800** [geobretagne/mviewer#130]
Bonjour,
J'essaie pour une couche customlayer d'appliquer la démo du cluster.js pour mon projet. Toutefois, en reprenant le code, j'ai réussi à générer la légende, mais rien ne s'affiche dans ma carte sinon (aucun lieu ni cluster). Voici mon codage pour la couche :
`                   {
              mviewer.customLayers.test_dist2 = {};
              var cl = mviewer.customLayers.test_dist2;
        cl.legend = { items:[] };
        var uniqueStyle = [
        new ol.style.Style({
            image: new ol.style.Circle({
                radius: 10,
                fill: new ol.style.Fill({
                    color: 'rgba(231, 76, 60, 0.7)'
                })
            })
        }),
        new ol.style.Style({
            image: new ol.style.Circle({
                radius: 8,
                fill: new ol.style.Fill({
                    color: 'rgba(236, 240, 241,7.0)'
                })
            })
        })
    ];
    cl.legend.items.push({styles:uniqueStyle, label: "Dossier", geometry: "Point"});
   var manyStyle = function (radius, radius2, size) {
        return [
            new ol.style.Style({
                image: new ol.style.Circle({
                    radius: radius,
                    fill: new ol.style.Fill({
                        color: 'rgba(236, 240, 241,0.7)'
                    })
                }),
                stroke: new ol.style.Stroke({
                    color: 'red',
                    width: 3
              }),
              fill: new ol.style.Fill({
                color: 'rgba(0, 0, 255, 0.1)'
              })
            }),
            new ol.style.Style({
                image: new ol.style.Circle({
                    radius: radius2,
                    fill: new ol.style.Fill({
                        color: 'rgba(231, 76, 60, 0.7)'
                    })
                }),
                text: new ol.style.Text({
                    font: '12px roboto_regular, Arial, Sans-serif',
                    text: size.toString(),
                    fill: new ol.style.Fill({
                        color: '#fff'
                    })
                })
            })
        ];
    };
    cl.legend.items.push({styles:manyStyle(10,10,7), label: "Groupe de dossiers", geometry: "Point"});

    var clusterStyle = function(feature) {
        var size = feature.get('features').length;
        var max_radius = 40;
        var max_value = 500;
        var radius = 10 + Math.sqrt(size)*(max_radius / Math.sqrt(max_value));
        var radius2 = radius *80 /100 ;
        if (size == 1) {
            return uniqueStyle;
        } else {
            return manyStyle(radius, radius2, size);
        }
    };

    cl.layer = new ol.layer.Vector({
        source: new ol.source.Cluster({
            distance: 50,
            source: new ol.source.Vector({
                url: "https://extranet.ville-houilles.fr/geoserver/houillesSIG/ows?service=WFS&version=1.0.0&request=GetFeature&typeName=houillesSIG:test_dist2&maxFeatures=50&outputFormat=application%2Fjson",
                format: new ol.format.GeoJSON()
            })
        }),
        style: clusterStyle

    });
    cl.handle = function(clusters, views) {
        if (clusters.length > 0 && clusters[0].properties.features) {
            var features = clusters[0].properties.features;
            var elements = [];
            var l = mviewer.getLayer("test_dist2");
            features.forEach(function(feature, i) {
                elements.push({
                    properties: feature.getProperties()
                });
            });
            var html;
            if (l.template) {
                html = info.templateHTMLContent(elements, l);
            } else {
                html = info.formatHTMLContent(elements, l);
            }
            var view = views["right-panel"];
            view.layers.push({
                "id": view.layers.length + 1,
                "firstlayer": true,
                "manyfeatures": (features.length > 1),
                "nbfeatures": features.length,
                "name": l.name,
                "layerid": "test_dist2",
                "theme_icon": l.icon,
                "html": html
            });
        }

    };
}`

Quelqu'un saurait me dire où se trouve mon erreur ? Merci beaucoup par avance de vos réponses !

**Comment by spelhate** [comment]
Il m'est difficile d'apporter une aide sans pouvoir disposer de l'exemple en ligne.
Quelques conseils cependant :

- Si rien ne s'affiche, il faut d'abord vérifier que la source de données est bien disponible et que la projection est la bonne. 
- Ensuite pour débugger un customlayer, il faut ajouter la ligne `debugger;` dans les endroits stratégiques du  code afin de pouvoir tracer le problème.

**Comment by fefe78800** [comment]
D'accord. L'exemple dont je parle est celui-ci : [http://kartenn.region-bretagne.fr/kartoviz/?config=demo/cluster.xml](url). Pour parler de l'affichage des données, je me suis permis de faire un test en remplaçant mes données personnelles par le lien url des données de la démo et la visualisation fonctionne (mais donc pas avec mes données de couche). 
Y a-t-il une manipulation spécifique à faire dans la saisie json des données (en faisant ce que je fais sur geoserver la manip de prévisualisation de couche en format geojson) ?

**Comment by spelhate** [comment]
Je parlais de votre exemple : https://extranet.ville-houilles.fr/mviewer/... ;-)

Je crois que la projection de vos données n'est pas la même que la projection de la démo.

**Comment by fefe78800** [comment]
Voici un lien vers le projet : [https://extranet.ville-houilles.fr/mviewer/chantiers_test.html#](url). J'ai tenté de changer la projection en ajoutant le suffixe derrière le lien mais sans résultat non plus

**Comment by spelhate** [comment]
Merci. Au lancement de l'application, la console  indique l' erreur suivante : 

> Assertion failed. See https://openlayers.org/en/v5.3.0/doc/errors/#10 for details.

D'après la doc et cette [ressource](https://github.com/openlayers/openlayers/issues/4814) , il semble que la fonction cluster ne fonctionne pas bien avec des géométries de type MultiPoint, ce qui est le cas dans votre exemple.

**Comment by fefe78800** [comment]
Bonjour,
Merci beaucoup ! En effet cela fonctionne en points et non en multipoints. En effectuant une conversion sur QGIS des entités multiples en entités simples, cela fonctionne correctement.

D'autre part, je vous avais parlé d'une documentation que je souhaitais vous faire part sur un détail de fonctionnalités, pouvant être intéressantes pour ceux qui voudraient. Elle constituera un premier jet, mais pourra être enrichie ultérieurement par moi-même si besoin. Où pourrais-je vous la déposer ?

**Comment by spelhate** [comment]
Sur Github ;-)

**Comment by fefe78800** [comment]
Bonjour,
Je tiens à préciser que j'ai réussi à réaliser le cluster simple, mais aussi une version catégorisée (avec 4 couleurs différentes au lieu de 2), et cette méthode sera prochainement ajoutée dans la documentation présente ici #138 .

**Comment by spelhate** [comment]
Merci !

**Comment by Gaetanbrl** [comment]
Le problème sur les clusters (cette issue) est-il résolu ?

**Comment by fefe78800** [comment]
Oui, l'erreur était liée au format des entités de la couche (multipoint au lieu de point). La suite devait être de faire une catégorisation attributaire d'éléments avec le cluster, chose faite également.
Je ferme le sujet si vous le souhaitez

**Comment by Gaetanbrl** [comment]
Super ! je pense donc que vous pouvez fermer l'issue.
